File: caesar-util/src/paths.rs
/* Copyright (C) 2024  Tomat et al.
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Affero General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later versions.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses>.
 *
 * Contents of this file are taken from Replugged, which is licensed under the
 * MIT License. The original source code can be found at:
 * <https://github.com/replugged-org/replugged/blob/main/scripts/inject/platforms>
 * The Replugged repositroy can be found at:
 * <https://github.com/replugged-org/replugged>
 * And the original license can be found at:
 * <https://github.com/replugged-org/replugged/blob/main/LICENSE>
 */

use std::{fmt::Display, path::PathBuf};

/// The "flavor" (or release) of Discord.
#[derive(Clone, Copy, Debug)]
pub enum DiscordFlavor {
    /// Stable, the main release of Discord.
    Stable,

    /// "Public Test Build", more stable than Canary but less stable than Stable.
    Ptb,

    /// Used for alpha testing, not as stable as PTB.
    Canary,

    /// Used for development, staging.
    Dev,
}

/// Handles locating Discord installations.
pub trait DiscordPathResolver {
    /// Resolves the paths to Discord installations.
    ///
    /// # Returns
    ///
    /// A list of Discord paths.
    fn resolve_discord_paths(&self) -> Vec<DiscordPath>;
}

impl DiscordPathResolver for DiscordFlavor {
    #[cfg(target_os = "windows")]
    fn resolve_discord_paths(&self) -> Vec<DiscordPath> {
        let folder = match self {
            DiscordFlavor::Stable => "Discord",
            DiscordFlavor::Ptb => "DiscordPTB",
            DiscordFlavor::Canary => "DiscordCanary",
            DiscordFlavor::Dev => "DiscordDevelopment",
        };
        let local_app_data = std::env::var("LOCALAPPDATA").unwrap();
        let discord_dir = PathBuf::from(local_app_data).join(folder);

        // TODO: Maybe return all of these potential paths?
        // ^ Would make old version filtering the burden of the API consumer.

        let target_app_dir = discord_dir.read_dir();
        if target_app_dir.is_err() {
            return vec![];
        }

        // Filter for directories that start with "app-", but reverse the order
        // to get the latest version.
        let target_app_dir = target_app_dir
            .unwrap()
            .map(|entry| entry.unwrap().path())
            .filter(|path| path.is_dir())
            .collect::<Vec<_>>()
            .into_iter()
            .rev()
            .find(|path| {
                path.file_name()
                    .unwrap()
                    .to_str()
                    .unwrap()
                    .starts_with("app-")
            })
            .unwrap();

        // This path is known to be constant.
        let asar_path = target_app_dir.join("resources").join("app.asar");

        // See earlier TODO regarding returning all paths.
        vec![DiscordPath {
            flavor: self.clone(),
            directory_path: target_app_dir,
            asar_path,
        }]
    }

    // TODO: Linux support.
    // TODO: Mac support.
    // I can implement both of these, but I don't have machines to test on and
    // I can't be asked.

    #[cfg(target_os = "linux")]
    fn resolve_discord_paths(&self) -> Vec<DiscordPath> {
        todo!()
    }

    #[cfg(target_os = "macos")]
    fn resolve_discord_paths(&self) -> Vec<DiscordPath> {
        todo!()
    }
}

impl Display for DiscordFlavor {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        // TODO: Are there official ways Discord presents these versions? I
        // ultimately don't care too much, but it may help with consistency.
        match self {
            DiscordFlavor::Stable => write!(f, "Discord Stable"),
            DiscordFlavor::Ptb => write!(f, "Discord PTB"),
            DiscordFlavor::Canary => write!(f, "Discord Canary"),
            DiscordFlavor::Dev => write!(f, "Discord Dev"),
        }
    }
}

/// A Discord installation path.
#[derive(Clone, Debug)]
pub struct DiscordPath {
    /// The flavor of Discord.
    pub flavor: DiscordFlavor,

    /// The directory path of the Discord installation.
    pub directory_path: PathBuf,

    /// The path to the `app.asar` file.
    pub asar_path: PathBuf,
}

impl Display for DiscordPath {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{} ({})", self.flavor, self.directory_path.display())
    }
}

/// Resolves the paths to Discord installations.
///
/// # Arguments
///
/// * `flavor` - The flavor of Discord to resolve paths for.
///
/// # Returns
///
/// A list of Discord paths.
pub fn resolve_discord_paths(flavor: DiscordFlavor) -> Vec<DiscordPath> {
    flavor.resolve_discord_paths()
}

File: discord/src/util/paths.ts
/* Copyright (C) 2024  Tomat et al.
 *
 * This program is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Affero General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later versions.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses>.
 */

import { basename, dirname, join } from "node:path";
import { lstatSync, mkdirSync, readdirSync, rmdirSync } from "node:fs";
import { app } from "electron";
import { getBuildInfo, type CaesarBuildInfo } from "./build-info";
import { log } from "./logging";

let resourcesPath: string | undefined;
let userDataPath: string | undefined;
let versionedUserDataPath: string | undefined;
let moduleDataPath: string | undefined;
let installPath: string | undefined;
let exeDir: string | undefined;

/**
 * Initializes paths.
 * @param buildInfo The build info to initialize within the context of.
 */
export function initializePaths(buildInfo: CaesarBuildInfo): void {
    if (!require.main)
        throw new Error("Cannot initialize paths without a main module");

    resourcesPath = join(require.main.filename, "..", "..", "..");
    userDataPath = findUserDataDir(findAppDataDir(), buildInfo);
    versionedUserDataPath = join(userDataPath, buildInfo.version);
    moduleDataPath =
        buildInfo.localModulesRoot ?? buildInfo.newUpdater
            ? join(userDataPath, "module_data")
            : join(versionedUserDataPath, "modules");

    exeDir = dirname(app.getPath("exe"));
    if (/^app-[0-9]+\.[0-9]+\.[0-9]+/.test(basename(exeDir))) {
        installPath = join(exeDir, "resources");
    }

    app.setPath("userData", userDataPath);
    mkdirSync(userDataPath, { recursive: true });

    log("paths", "resourcesPath:", resourcesPath);
    log("paths", "userDataPath:", userDataPath);
    log("paths", "userDataVersionedPath:", versionedUserDataPath);
    log("paths", "moduleDataPath:", moduleDataPath);
    log("paths", "installPath:", installPath);
    log("paths", "exeDir:", exeDir);
}

/**
 * Removes old (outdated) Discord installations.
 * @param buildInfo The build info.
 */
export function removeOldInstallations(buildInfo: CaesarBuildInfo): void {
    if (!userDataPath) {
        log("paths", "userDataPath not set, cannot remove old installations");
        return;
    }

    log("paths", "Removing old installations");

    const entries = readdirSync(userDataPath);
    for (const entry of entries) {
        if (!userDataPath) {
            return;
        }

        const fullPath = join(userDataPath, entry);
        try {
            const stat = lstatSync(fullPath);

            if (!stat.isDirectory()) {
                return;
            }

            if (entry.includes(buildInfo.version)) {
                return;
            }

            if (!/^[0-9]+.[0-9]+.[0-9]+/.exec(entry)) {
                return;
            }

            log("paths", "Found old installation, removing", fullPath);

            rmdirSync(fullPath, { recursive: true });
        } catch (e) {
            log("paths", "Failed to remove old installation", fullPath, e);
            return;
        }
    }
}

/**
 * @returns The path to which Discord is installed.
 */
export function getInstallPath(): string | undefined {
    return installPath;
}

/**
 * @returns The path to which module data is saved.
 */
export function getModuleDataPath(): string | undefined {
    return moduleDataPath;
}

/**
 * @returns The Electron `resources` directory.
 */
export function getResourcesPath(): string | undefined {
    return resourcesPath;
}

/**
 * @returns The user data `discord{releaseChannel}` directory.
 */
export function getUserDataPath(): string | undefined {
    return userDataPath;
}

/**
 * @returns The versioned sub-directory beneath the user data directory.
 */
export function getVersionedUserDataPath(): string | undefined {
    return versionedUserDataPath;
}

export function getExecutableDir(): string | undefined {
    return exeDir;
}

function findAppDataDir(): string {
    userDataPath = process.env.DISCORD_USER_DATA_DIR;
    if (userDataPath) {
        return userDataPath;
    }

    return app.getPath("appData");
}

function findUserDataDir(
    userDataRoot: string,
    buildInfo: CaesarBuildInfo,
): string {
    const appName = getBuildInfo().caesarStandalone ? "caesar" : "discord";
    return join(
        userDataRoot,
        appName +
            (buildInfo.releaseChannel === "stable"
                ? ""
                : buildInfo.releaseChannel),
    );
}
